### PR TITLE
Make search more consistent

### PR DESCRIPTION
**Pull request by bkamins**
This is an implementation of #24103. If will fail the tests now. I will fix the tests when there is an agreement that the proposed functionality is OK.
All key assumptions how `search` now works are given in its docstring

### CODE REVIEW

File: base/strings/search.jl [hunk 1]
```diff
@@ -3,18 +3,39 @@
...
-    search(string::AbstractString, chars::Chars, [start::Integer])
+    search(string, chars, [start::Integer])
 
 Search for the first occurrence of the given characters within the given string. The second
 argument may be a single character, a vector or a set of characters, a string, or a regular
 expression (though regular expressions are only allowed on contiguous strings, such as ASCII
-or UTF-8 strings). The third argument optionally specifies a starting index. The return
-value is a range of indexes where the matching sequence is found, such that `s[search(s,x)] == x`:
+or UTF-8 strings).
+Alternatively the first and the second arguments may be arrays of `Int8` or `UInt8`.
```
Comment (nalimilan) [inline base/strings/search.jl]: I'd rather not document this, it really should be an internal function which shouldn't have been exposed. I was preparing a PR to merge `search` with `findfirst` which would also handle this.

File: base/strings/search.jl [hunk 2]
```diff
@@ -3,18 +3,39 @@
...
+The third argument optionally specifies a starting index.
+If it is not a valid index into the first argument then error is thrown.
```
Comment (nalimilan) [inline base/strings/search.jl]: This goes without saying, I don't think it's worth making the description longer than it already is to mention this.

File: base/strings/search.jl [hunk 3]
```diff
@@ -3,18 +3,39 @@
...
+The return value when the second argument is a string or a vector of bytes is the samllest
```
Comment (nalimilan) [inline base/strings/search.jl]: "samllest". "exits".

File: base/strings/search.jl [hunk 4]
```diff
@@ -3,18 +3,39 @@
...
-`0:-1` if unmatched.
+`search(string, "substring", i)` = `start:end` such that `string[start:end] == "substring"`
+and `start >= i`.
 
-`search(string, 'c')` = `index` such that `string[index] == 'c'`, or `0` if unmatched.
+In particular a search for a zero length second argument returns `i:(i-1)`.
+
+If the first argument is empty and third argument is not given `0:-1` is returned.
+
+The return value when the second argument is a signle character, a vector or a set of
```
Comment (nalimilan) [inline base/strings/search.jl]: "signle"

File: base/strings/search.jl [hunk 5]
```diff
@@ -3,18 +3,39 @@
...
 
-`search(string, "substring")` = `start:end` such that `string[start:end] == "substring"`, or
-`0:-1` if unmatched.
+`search(string, "substring", i)` = `start:end` such that `string[start:end] == "substring"`
+and `start >= i`.
 
-`search(string, 'c')` = `index` such that `string[index] == 'c'`, or `0` if unmatched.
+In particular a search for a zero length second argument returns `i:(i-1)`.
+
+If the first argument is empty and third argument is not given `0:-1` is returned.
```
Comment (nalimilan) [inline base/strings/search.jl]: Better give the general rule and use `start=start(s)` in the signature. People will figure out what happens in special cases.

File: base/strings/search.jl [hunk 6]
```diff
@@ -23,176 +44,190 @@ julia> search("Hello to the world", "z")
...
         i = j
     end
     return 0
 end
-search(s::AbstractString, c::Chars) = search(s,c,start(s))
 
-in(c::Char, s::AbstractString) = (search(s,c)!=0)
+search(s::AbstractString, c::Chars) = s == "" ? 0 : search(s, c, start(s))
+
+in(c::Char, s::AbstractString) = search(s,c) != 0
```
Comment (nalimilan) [inline base/strings/search.jl]: Plase don't change unrelated lines. Same below. This makes it really painful to review.

File: base/strings/search.jl [hunk 7]
```diff
@@ -23,176 +44,190 @@ julia> search("Hello to the world", "z")
...
 
-_searchindex(s, t::Char, i) = search(s, t, i)
-
 function _search_bloom_mask(c)
     UInt64(1) << (c & 63)
 end
 
-_nthbyte(s::String, i) = codeunit(s, i)
-_nthbyte(a::ByteArray, i) = a[i]
+_nthbyte(s::String, i) = @inbounds codeunit(s, i)
```
Comment (nalimilan) [inline base/strings/search.jl]: Better add `@inbounds` at the call site, and add `@propagate_inbounds` annotations here.

File: base/strings/search.jl [hunk 8]
```diff
@@ -23,176 +44,190 @@ julia> search("Hello to the world", "z")
...
 
-searchindex(s::ByteArray, t::ByteArray, i) = _searchindex(s,t,i)
-
 """
-    searchindex(s::AbstractString, substring, [start::Integer])
+    searchindex(s, t, [start::Integer])
 
-Similar to [`search`](@ref), but return only the start index at which
-the substring is found, or `0` if it is not.
+Similar to [`search`](@ref), but return only the start index at which the second argument
```
Comment (nalimilan) [inline base/strings/search.jl]: "substring" was clearer, let's not make the docs worse because of that `UInt8` thing.

**Review by nalimilan — COMMENTED** [review]
There are too many unrelated formatting changes in this PR. Also I won't repeat comments I made at https://github.com/JuliaLang/julia/pull/24116. I'll review it again when it's more focused.

**Comment by bkamins** [comment]
@nalimilan Sorry for messing with the code formatting - lesson taken and I will fix it.

Before I correct this PR I would like a clear recommendation about the issue also discussed in #24116 regarding `start` argument of `search`.

Currently:
```
julia> search("", "asd")
0:-1
```
and I think this is what users would expect.
On the other hand:
```
julia> search("", "asd", 10)
0:-1
```
is something we want to avoid as `10` is an invalid index.

The only problem is what we want to return in `search("", "asd")` vs. `search("", "asd", 1)`.
My reasoning is that the second form should throw an error. And I assume we have a consensus here as `1` is an invalid index for `""`. However, I think that `search("", "asd")` should still return `0:-1`. Otherwise users will have to write their code as something like:
```
if endof(s) >= start(s)
    x = search(s, "some string")
    # do something with x
else
    # do the same thing knowing that no match was found
end
```
which is painful and error prone as in general user might not know if `s` is empty.

Remember that my proposal additionally fixes:
```
julia> search("","")
1:0
```
which is incorrect for sure. But I would say that `0:-1` should be returned (and not an error thrown in this case).

@nalimilan What do you think is a proper way to go given the above?

**Comment by nalimilan** [comment]
I hadn't realized it was such a Catch 22. I'd say we should have a special case for `1` with `""`, since `start("") == 1`, and `start(s)` is the default value for the index.

The only other solution I can think of is to allow `search("", "asd")` but not `search("", "asd", 1)`, but that would be cumbersome to implement and problematic for users in they want to write generic code with an index falling back to `start(s)`. So making an exception sounds better.

**Comment by bkamins** [comment]
Just to make sure I understand your sentence:
> So making an exception sounds better.
You are opting to make search on `""` always throw an exception (when searching for substrings, regexps and `Chars`)?

If this is the decision I can implement it and it should go into the documentation then.

**Comment by nalimilan** [comment]
That's my preferred solution at this stage, but you may want to way for possible comments at https://github.com/JuliaLang/julia/issues/24103#issuecomment-336814417.

**Comment by bkamins** [comment]
closing as it is outdated.
